pkg/clusteragent/autoscaling/workload/local/replica_calculator.go

```diff
@@ -0,0 +1,325 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2024-present Datadog, Inc.
+
+//go:build kubeapiserver
+
+// Package local provides local recommendations for autoscaling workloads.
+package local
+
+import (
+	"fmt"
+	"math"
+	"time"
+
+	datadoghq "github.com/DataDog/datadog-operator/api/datadoghq/v1alpha1"
+
+	workloadmeta "github.com/DataDog/datadog-agent/comp/core/workloadmeta/def"
+
+	"github.com/DataDog/datadog-agent/pkg/clusteragent/autoscaling/workload/common"
+	"github.com/DataDog/datadog-agent/pkg/clusteragent/autoscaling/workload/loadstore"
+	"github.com/DataDog/datadog-agent/pkg/clusteragent/autoscaling/workload/model"
+	le "github.com/DataDog/datadog-agent/pkg/util/kubernetes/apiserver/leaderelection/metrics"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+)
+
+type replicaCalculator struct {
+	PodWatcher common.PodWatcher
+}
+
+type utilizationResult struct {
+	averageUtilization      float64
+	podToUtilization        map[string]float64
+	missingPods             []string
+	recommendationTimestamp time.Time
+}
+
+func newReplicaCalculator(podWatcher common.PodWatcher) replicaCalculator {
+	return replicaCalculator{
+		PodWatcher: podWatcher,
+	}
+}
+
+// CalculateHorizontalRecommendations is the entrypoint to calculate the horizontal recommendation for a given DatadogPodAutoscaler
+func (r replicaCalculator) CalculateHorizontalRecommendations(dpai model.PodAutoscalerInternal, lStore loadstore.Store) (*model.ScalingValues, error) {
+	currentTime := time.Now()
+
+	// Get current pods for the target
+	targetRef := dpai.Spec().TargetRef
+	targets := dpai.Spec().Targets
+	targetGVK, targetErr := dpai.TargetGVK()
+	if targetErr != nil {
+		return nil, fmt.Errorf("Failed to get GVK for target: %s, %s", dpai.ID(), targetErr)
+	}
+
+	podOwnerName := targetRef.Name
+	namespace := dpai.Namespace()
+
+	podOwner := common.NamespacedPodOwner{
+		Namespace: namespace,
+		Name:      podOwnerName,
+		Kind:      targetGVK.Kind,
+	}
+	pods := r.PodWatcher.GetPodsForOwner(podOwner)
+	if len(pods) == 0 {
+		// If we found nothing, we'll wait just until the next sync
+		return nil, fmt.Errorf("No pods found for autoscaler: %s, gvk: %s, name: %s", dpai.ID(), targetGVK.String(), targetRef.Name)
+	}
+
+	recommendedReplicas := model.HorizontalScalingValues{}
+
+	for _, target := range targets {
+		recSettings, err := newResourceRecommenderSettings(target)
+		if err != nil {
+			return nil, fmt.Errorf("Failed to get resource recommender settings: %s", err)
+		}
+
+		queryResult := lStore.GetMetricsRaw(recSettings.metricName, namespace, podOwnerName, recSettings.containerName)
+		rec, utilizationRes, err := recSettings.recommend(currentTime, pods, queryResult, float64(*dpai.CurrentReplicas()))
```

Comment:
`dpai.CurrentReplicas()` may return nil. It's a race with the Horizontal controller, especially on a newly created DPA.
I'm also thinking that you could get this info from `PodWatcher` (to be consistent with the number of ready pods)